collect_app/src/main/java/org/odk/collect/android/tasks/InstanceUploaderTask.java

```diff
@@ -490,8 +496,14 @@ private boolean uploadOneSubmission(String urlString, String id, String instance
                         outcome.mResults.put(id, fail + response.getStatusLine().getReasonPhrase()
                                 + " (" + responseCode + ") at " + urlString);
                     } else {
-                        outcome.mResults.put(id, fail + response.getStatusLine().getReasonPhrase()
-                                + " (" + responseCode + ") at " + urlString);
+                        // If response from server is valid use that else use default messaging
+                        if (messageParser != null && messageParser.isValid()){
```

Comment:
It's not possible for it to be null because it gets initialized on line 484 which is always guaranteed to execute before you get here. Plus, it's always guaranteed to run before you get to line 529 because of the `|| first` condition which guarantees the while loop is always run at least once.

Line 340 (`messageParser = null;`) is misleading. In fact, it shouldn't be there at all because its value is always overwritten. Line 484 is in a try block but actually it's at the same level as the assignment to null and always guaranteed to run after it.

Now I'm noticing that Android Studio does perform this analysis for you and lets you know that line 340 can be removed and that the null checks are always guaranteed to be true! Do you have those checks turned on?